registrations/permissions.py

```diff
@@ -1,46 +1,91 @@
-from django.utils.translation import gettext as _
 from rest_framework import permissions
 from rest_framework.request import Request
 from rest_framework.views import APIView
 
 from events.auth import ApiKeyUser
-from registrations.models import SignUp
+from events.permissions import UserDataFromRequestMixin
+from registrations.models import Registration, SignUp
 
 
-class CanCreateEditDeleteSignup(permissions.BasePermission):
-    message: str = _(
-        "Only authenticated users are able to access sign-ups. Viewing, editing and deleting are"
-        "allowed only for admins of the publishing organization and for users that have created "
-        "the sign-up."
-    )
-
+class CanAccessRegistration(UserDataFromRequestMixin, permissions.BasePermission):
     def has_permission(self, request: Request, view: APIView) -> bool:
-        return (
-            request.method
-            in permissions.SAFE_METHODS + ("POST", "PUT", "PATCH", "DELETE")
-            and request.user.is_authenticated
-        )
+        if not request.user.is_authenticated:
+            return False
+
+        if request.method == "POST":
+            (
+                __,
+                user_organization,
+            ) = self.user_data_source_and_organization_from_request(request)
+            return (
+                request.user.is_superuser
+                or request.user.is_admin_of(user_organization)
+                or request.user.is_registration_admin_of(user_organization)
+            )
+
+        if request.method in ("PUT", "PATCH", "DELETE"):
+            (
+                __,
+                user_organization,
+            ) = self.user_data_source_and_organization_from_request(request)
+            return bool(user_organization)
+
+        return request.method in permissions.SAFE_METHODS
 
     def has_object_permission(
-        self, request: Request, view: APIView, obj: SignUp
+        self, request: Request, view: APIView, obj: Registration
     ) -> bool:
         if isinstance(request.user, ApiKeyUser):
             user_data_source, _ = view.user_data_source_and_organization
             # allow only if the api key matches instance data source
             if obj.data_source != user_data_source:
                 return False
 
+        if request.method == "GET":
+            return (
+                obj.can_be_edited_by(request.user)
+                or obj.registration_user_accesses.filter(
```

Comment:
Minor risk of privilege escalation if the user is able to manipulate their email address. But maybe nothing to worry about considering the use case? 

Minimizing the risk would probably need something that binds the RegistrationUserAccess permanently to a User object (as soon as such object becomes available).

Reply:
Yeah I don't see any easy way spoofing email with the current authentication methods in tunnistamo.